eth_utils/abi.py

```diff
@@ -1,68 +1,784 @@
+    :rtype: `str`
+
+    .. doctest::
+
+        >>> from eth_utils.abi import collapse_if_tuple
+        >>> abi = {
+        ...   'components': [
+        ...     {'name': 'anAddress', 'type': 'address'},
+        ...     {'name': 'anInt', 'type': 'uint256'},
+        ...     {'name': 'someBytes', 'type': 'bytes'},
+        ...   ],
+        ...   'type': 'tuple',
+        ... }
+        >>> collapse_if_tuple(abi)
+        '(address,uint256,bytes)'
+    """
+    if isinstance(abi, str):
+        return abi
+
+    element_type = abi.get("type")
+    if not isinstance(element_type, str):
         raise TypeError(
-            f"The 'type' must be a string, but got {repr(typ)} of type {type(typ)}"
+            f"The 'type' must be a string, but got {repr(element_type)} of type "
+            f"{type(element_type)}"
         )
-    elif not typ.startswith("tuple"):
-        return typ
+    elif not element_type.startswith("tuple"):
+        return element_type
 
     delimited = ",".join(collapse_if_tuple(c) for c in abi["components"])
     # Whatever comes after "tuple" is the array dims. The ABI spec states that
     # this will have the form "", "[]", or "[k]".
-    array_dim = typ[5:]
+    array_dim = element_type[5:]
     collapsed = f"({delimited}){array_dim}"
 
     return collapsed
 
 
-def _abi_to_signature(abi: Dict[str, Any]) -> str:
-    fn_input_types = ",".join(
-        [collapse_if_tuple(abi_input) for abi_input in abi.get("inputs", [])]
+def abi_to_signature(abi_element: ABIElement) -> str:
+    """
+    Returns a string signature representation of the function or event ABI
+    and arguments.
+
+    Signatures consist of the name followed by a list of arguments.
+
+    :param abi_element: ABI element.
+    :type abi_element: `ABIElement`
+    :return: Stringified ABI signature
+    :rtype: `str`
+
+    .. doctest::
+
+        >>> from eth_utils import abi_to_signature
+        >>> abi_element = {
+        ...   'constant': False,
+        ...   'inputs': [
+        ...     {
+        ...       'name': 's',
+        ...       'type': 'uint256'
+        ...     }
+        ...   ],
+        ...   'name': 'f',
+        ...   'outputs': [],
+        ...   'payable': False,
+        ...   'stateMutability': 'nonpayable',
+        ...   'type': 'function'
+        ... }
+        >>> abi_to_signature(abi_element)
+        'f(uint256)'
+    """
+    signature = "{name}({input_types})"
+
+    abi_type = str(abi_element.get("type", ""))
+    if abi_type == "fallback" or abi_type == "receive":
+        return signature.format(name=abi_type, input_types="")
+
+    if abi_type == "constructor":
+        fn_name = abi_type
+    else:
+        fn_name = str(abi_element.get("name", abi_type))
+
+    return signature.format(
+        name=fn_name, input_types=",".join(get_abi_input_types(abi_element))
+    )
+
+
+def filter_abi_by_name(name: str, contract_abi: ABI) -> Sequence[ABIElement]:
```

Comment:
This is definitely a nit, but see my comment in `filter_abi_by_type()`. I think it'd be best to be consistent with the naming and rename this to `abi_name`.


```suggestion
def filter_abi_by_name(abi_name: str, contract_abi: ABI) -> Sequence[ABIElement]:
```

I won't add separate comments about the docstring here, but we'll need to update the docstring names as well.